### PR TITLE
Indent guides

### PR DESCRIPTION
**Pull request by bennetbo**
Builds on top of existing work from #2249, but here's a showcase:

https://github.com/zed-industries/zed/assets/53836821/4b346965-6654-496c-b379-75425d9b493f

TODO:
- [x] handle line wrapping
- [x] implement handling in multibuffer (crashes currently)
- [x] add configuration option
- [x] new theme properties? What colors to use?
- [x] Possibly support indents with different colors or background colors
- [x] investigate edge cases (e.g. indent guides and folds continue on empty lines even if the next indent is different)
- [x] add more tests (also test `find_active_indent_index`)
- [x] docs (will do in a follow up PR)
- [x] benchmark performance impact

Release Notes:

- Added indent guides ([#5373](https://github.com/zed-industries/zed/issues/5373))

### CODE REVIEW

**Comment by d1y** [comment]
It looks great, I hope it can have built-in rainbow indentation :)

**Comment by bennetbo** [comment]
> It looks great, I hope it can have built-in rainbow indentation :)

@d1y It is actually built-in. In your settings file you will be able to do something like this:

```json
"experimental.theme_overrides": {
    "accents": [ // Specify some accent colors to use for the indent guide lines/background
      "#ffff40",
      "#7fff7f",
      "#ff7fff",
      "#4fecec",
      "#ffff40",
      "#7fff7f",
      "#ff7fff",
      "#4fecec",
      "#ffff40",
      "#7fff7f",
      "#ff7fff",
      "#4fecec"
    ]
  },
  "indent_guides": {
    "coloring": "indent_aware",  // Set this to indent_aware to paint the lines in rainbow colors
    "background_coloring": "indent_aware" // Set this to indent_aware to paint the background in rainbow colors
  },
```

You can also only colorize/disable the lines/the background by customizing the `coloring` and `background_coloring` values

image

**Comment by baldwindavid** [comment]
@bennetbo Great work! For the "fixed" color mode is there a way to set the color of the active and inactive indent guide via the theme? I see an example for the gruvbox theme accents, but thinking that is only for "indent_aware" colors. What I would be shooting for is to set the "active" indent guide to a more prominent color and the inactive guides to transparent. Regardless of whether that is in here this will be a great addition. Thanks!

**Comment by baldwindavid** [comment]
The other notes for future consideration are:
- It would be nice to be able to toggle on and off with a command like `editor::ToggleIndentGuides` (similar to `editor::ToggleLineNumbers`).
- Some languages tend need indent guides more than others. For example, I don't typically want guides in Elixir, but need them in HTML-ish code with deeper nesting. Might be nice to support language overrides similar to those for hard warp, tab size, etc.

I'm not even sure I'd want to mess with the default active/inactive theme colors mentioned in my previous comment if these sorts of settings were available.

File: crates/language/src/buffer.rs [hunk 1]
```diff
@@ -3058,6 +3089,241 @@ impl BufferSnapshot {
...
+                range.end = Point::new(indent.end_row, 0).to_offset(self);
+            }
+        }
+
+        result_vec.extend(indent_stack);
+
+        result_vec
+    }
+
+    pub async fn enclosing_indent(&self, mut buffer_row: u32) -> Option<(Range<u32>, u32)> {
```
Comment (SomeoneToIgnore) [inline crates/language/src/buffer.rs:3205]: > u32, u32, 32

Passing by in curiosity, and wanted to note that using `MultiBufferRow` and `DisplayRow` might fix panics and reduce bugs.
Would be nice to remove all u32's possible around, but not insisting.

**Comment by bennetbo** [comment]
@baldwindavid Indent guides use the following theme color's `editor.indent_guide` and `editor.indent_guide_active` when `coloring` is set to `fixed` (which is the default), so they can be styled by the theme and also customized with theme overrides.
Thanks for the suggestions! Will add them in a follow up PR

**Comment by baldwindavid** [comment]
@bennetbo Super cool. I added `"editor.indent_guide": "#00000000"` and it works great.

I also see that you went ahead and added the per language config. I tried it out and quickly realized that my plan to disable for Elixir is kind of foiled in that this also disables it for heex (the elixir HTML template language). I tried this, but didn't seem to achieve what I was trying to do as the guides are still disabled for heex...

```json
  "languages": {
    "Elixir": {
      "language_servers": ["lexical", "!elixir-ls", "!next-ls", "..."],
      "indent_guides": {
        "enabled": false
      }
    },
    "HEEX": {
      "indent_guides": {
        "enabled": true
      }
    },
```

At any rate, don't think that has anything to do with this PR as the language-level settings do work. Thanks!

**Comment by bennetbo** [comment]
> @bennetbo Super cool. I added `"editor.indent_guide": "#00000000"` and it works great.
> 
> I also see that you went ahead and added the per language config. I tried it out and quickly realized that my plan to disable for Elixir is kind of foiled in that this also disables it for heex (the elixir HTML template language). I tried this, but didn't seem to achieve what I was trying to do as the guides are still disabled for heex...
> 
> ```json
>   "languages": {
>     "Elixir": {
>       "language_servers": ["lexical", "!elixir-ls", "!next-ls", "..."],
>       "indent_guides": {
>         "enabled": false
>       }
>     },
>     "HEEX": {
>       "indent_guides": {
>         "enabled": true
>       }
>     },
> ```
> 
> At any rate, don't think that has anything to do with this PR as the language-level settings do work. Thanks!

I am not familiar with HEEX and elixir, is the HEEX code in the same file as the elixir code? Right now there's no support for rendering indent guides only for a certain part of a file (the language settings get inferred from the first line of the buffer, which in your case is probably elixir -> that's why they don't show up)

**Comment by baldwindavid** [comment]
Elixir is the main language and HEEX a templating language for Elixir. They are both in the [elixir extension within the zed repo](https://github.com/zed-industries/zed/tree/main/extensions/elixir/languages).

While the elixir extension does support heex injections, the scenario I'm talking about here is for elixir in a file with a `.ex` extension and a heex template in a different file with the extension `.heex`.

**Comment by bennetbo** [comment]
> Elixir is the main language and HEEX a templating language for Elixir. They are both in the [elixir extension within the zed repo](https://github.com/zed-industries/zed/tree/main/extensions/elixir/languages).
> 
> While the elixir extension does support heex injections, the scenario I'm talking about here is for elixir in a file with a `.ex` extension and a heex template in a different file with the extension `.heex`.

Hmm that is odd, I just tested it and specifying something like tab_size will also not work, HEEX always seems to use the settings that I specify for elixir
